pkg/templates/kubeadm/v1beta4/kubeadm.go

```diff
@@ -0,0 +1,621 @@
+		if !s.Cluster.CloudProvider.External {
+			switch {
+			case s.Cluster.ClusterNetwork.IPFamily == kubeoneapi.IPFamilyIPv4IPv6:
+				kubeletCLIFlags = setAllArgsValue(kubeletCLIFlags, "node-ip", newNodeIP(host)+","+host.IPv6Addresses[0])
+			case s.Cluster.ClusterNetwork.IPFamily == kubeoneapi.IPFamilyIPv6IPv4:
+				kubeletCLIFlags = setAllArgsValue(kubeletCLIFlags, "node-ip", host.IPv6Addresses[0]+","+newNodeIP(host))
+			}
+		}
+	} else {
+		kubeletCLIFlags = setAllArgsValue(kubeletCLIFlags, "node-ip", newNodeIP(host))
+	}
+
+	if m := host.Kubelet.SystemReserved; m != nil {
+		kubeletCLIFlags = setAllArgsValue(kubeletCLIFlags, "system-reserved", kubeoneapi.MapStringStringToString(m, "="))
+	}
+
+	if m := host.Kubelet.KubeReserved; m != nil {
+		kubeletCLIFlags = setAllArgsValue(kubeletCLIFlags, "kube-reserved", kubeoneapi.MapStringStringToString(m, "="))
+	}
+
+	if m := host.Kubelet.EvictionHard; m != nil {
+		kubeletCLIFlags = setAllArgsValue(kubeletCLIFlags, "eviction-hard", kubeoneapi.MapStringStringToString(m, "<"))
+	}
+	if m := host.Kubelet.MaxPods; m != nil {
+		kubeletCLIFlags = setAllArgsValue(kubeletCLIFlags, "max-pods", strconv.Itoa(int(*m)))
+	}
+
+	if m := host.Kubelet.PodPidsLimit; m != nil {
+		kubeletCLIFlags = setAllArgsValue(kubeletCLIFlags, "pod-max-pids", strconv.Itoa(int(*m)))
+	} else {
+		// Set default value if PodsPidsLimits is nil
+		// in order to pass the 4.2.13 Check in CIS Benchmark 1.8
+		kubeletCLIFlags = setAllArgsValue(kubeletCLIFlags, "pod-max-pids", strconv.Itoa(-1))
+	}
+
+	return kubeadmv1beta4.NodeRegistrationOptions{
+		Name:             host.Hostname,
+		Taints:           host.Taints,
+		CRISocket:        fmt.Sprintf("unix://%s", s.Cluster.ContainerRuntime.CRISocket()),
+		KubeletExtraArgs: kubeletCLIFlags,
+	}
+}
+
+func setAllArgsValue(args []kubeadmv1beta4.Arg, name, value string) []kubeadmv1beta4.Arg {
+	return kubeadmv1beta4.SetArgValues(args, name, value, -1)
+}
+
+func stringStringMapToArgs(m map[string]string) []kubeadmv1beta4.Arg {
+	args := []kubeadmv1beta4.Arg{}
+	for k, v := range m {
+		args = append(args, kubeadmv1beta4.Arg{
+			Name:  k,
+			Value: v,
+		})
+	}
+
+	return args
+}
+
+func argsToStringStringMap(args []kubeadmv1beta4.Arg) map[string]string {
+	m := map[string]string{}
+	for _, arg := range args {
+		m[arg.Name] = arg.Value
+	}
+
+	return m
+}
+
+func newNodeIP(host kubeoneapi.HostConfig) string {
+	nodeIP := host.PrivateAddress
+	if nodeIP == "" {
+		nodeIP = host.PublicAddress
+	}
+
+	return nodeIP
+}
+
+func join(ipFamily kubeoneapi.IPFamily, ipv4Subnet, ipv6Subnet string) string {
+	switch ipFamily {
+	case kubeoneapi.IPFamilyIPv4:
+		return ipv4Subnet
+	case kubeoneapi.IPFamilyIPv6:
+		return ipv6Subnet
+	case kubeoneapi.IPFamilyIPv4IPv6:
+		return strings.Join([]string{ipv4Subnet, ipv6Subnet}, ",")
+	case kubeoneapi.IPFamilyIPv6IPv4:
+		return strings.Join([]string{ipv6Subnet, ipv4Subnet}, ",")
+	default:
+		return "unknown IP family"
```

Comment:
This should be a panic(), no?

Reply:
No, it should return an error, panic doesn't fail nicely and we have pretty neat error handling for this. Shouldn't even be possible though, more like fail-safe mechanism.

Reply:
So now instead of doing either of the two correct things, we continue to just return a wrong value and continue on the happy path?

I understand that the IPFamily type should never ever have a value outside of the 4 that are listed. Which is exactly why the default case should be a panic, because there is something deeply wrong with the code if we hit that default case.

Reply:
https://github.com/kubermatic/kubeone/pull/3360